### ISSUE TITLE
ZeroTier L1 transport should only use RFC4193 addresses to other RFC4193 endpoints

### ISSUE THREAD

**Issue by cwichura** [zerotier/ZeroTierOne#670]
We have IPv6 configured to give internal hosts two IPv6 prefixes.  One prefix is our regular, public, globally routable prefix.  The second prefix is an RFC4193 unique local internal to the organization, and used to access a couple restricted networks that only have the RFC4193 prefix and NO public prefix.

Normally, IPv6's source address selection algorithm results in only using the RFC4193 as the source address when talking to another RFC4193 host.  (This has been working as expected for us since we implemented the dual prefixes.)  However, we have observed that ZeroTier attempts to use these RFC4193 addresses to talk to non-RFC4193 addresses, such as the ZeroTier root servers.  I assume this is because ZeroTier builds a list of all IPv6 addresses on the system and then iterates over all of them, explicitly sending its discovery/treejoin/whateveryoucallthem packets to systems it is attempting to build a L1 transport link with (be it the public roots, network controller, or other nodes within the private ZeroTier network), thus bypassing the OS' normal source address selection logic.

We have network ACLs that block all RFC4193-to-non-RFC4193 (and vice versa), and since we have started testing ZeroTier this past week, the drop counts on these rules have exploded.  Turning on logging for these rules, the drops are all ZeroTier's UDP 9993.

TL;DR: ZeroTier should only use RFC4193 as the source address of its L1 transport packets when the destination address is also an RFC4193 address.

**Comment by neutronscott** [comment]
This isn't just IPv6. ZeroTier is very aggressive with discovery. My home router is ARPing for zerotier nodes all day out the LAN interfaces:

10:51:31.996361 ARP, Request who-has 159.203.2.154 tell 192.168.2.1, length 28
10:51:31.996705 ARP, Request who-has 107.170.197.14 tell 192.168.1.1, length 28

The manual mentions you can blacklist physical paths in local.conf, so I added 192.168.0.0/16 to this, thinking it'd never use these addresses. It will still use them as source. Maybe it won't use them as destination.

Perhaps the proper fix is to make this also check source addresses. So the iteration done on remoteAddr here, can also be done on localAddr?

https://github.com/zerotier/ZeroTierOne/blob/fe5257df81c4ec4b5d48f707eb794de0748b7ac0/service/OneService.cpp#L2213

**Comment by adamierymenko** [comment]
The problem here is that some networks including Azure implement IPv6 NAT. Yes this is absolutely brain-damaged and evil and wrong and those responsible should be disemboweled as a warning to others, but it's a fact. On the networks private IPv6 addresses must be used to talk to everything. Blocking this would make ZeroTier stop working out of the box on these networks.

Blacklists are fixed in `dev` now which is about to be 1.2.6 and could be used to blacklist these IPs.

**Comment by adamierymenko** [comment]
In general to make ZeroTier work everywhere it must be pretty aggressive and "defensive," assuming that any interface might be the right one and never giving up.
